### PR TITLE
Fixed exploits with unexpected data, which causes the ServerRepositor…

### PR DESCRIPTION
**Pull request by SkippsDev**
…y to crash.

### CODE REVIEW

File: direct/src/distributed/ServerRepository.py [hunk 1]
```diff
@@ -372,20 +374,33 @@ def handleDatagram(self, datagram):
...
+        del self.clientsByDoIdBase[client.doIdBase]
+        self.qcr.removeConnection(connection)
 
     def handleClientCreateObject(self, datagram, dgi):
         """ client wants to create an object, so we store appropriate
         data, and then pass message along to corresponding zones """
 
         connection = datagram.getConnection()
+        
+        if dgi.getRemainingSize() < 3:
```
Comment (loblao) [inline direct/src/distributed/ServerRepository.py]: Still exploitable. Since `getUint32` reads 4 bytes and `getUint16` reads 2, you need 10 bytes.

**Comment by loblao** [comment]
There's a major issue in SR: it doesn't inspect client updates (`dg.appendData(dgi.getRemainingBytes())`) so those bytes are sent to the clients and could be manipulated to crash them.

File: direct/src/distributed/ServerRepository.py (hunk 1 above)
Comment (SkippsDev) [inline direct/src/distributed/ServerRepository.py]: Oh yeah, idek why i didn't see that... ugh....

**Comment by rdb** [comment]
Closing due to inactivity despite outstanding issues.

I feel that addressing this kind of issue requires a more thorough approach to handling out-of-bounds datagram reads.  Hardcoding expected byte lengths is a rather fragile approach.

**Pull request closed without merging**